### PR TITLE
deps: move libp2p-server and metrics example to use axum 0.7

### PR DESCRIPTION
**Pull request by drHuangMHT**
## Description

Move from `hyper` to `axum` for `libp2p-server` and metrics example.
Running in parallel with #5038.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by jxs** [comment]
Hi @drHuangMHT did you see https://github.com/libp2p/rust-libp2p/pull/5038?

**Comment by drHuangMHT** [comment]
> Hi @drHuangMHT did you see https://github.com/libp2p/rust-libp2p/pull/5038?

Oopsie, didn't see that.

**Comment by drHuangMHT** [comment]
Following https://github.com/libp2p/rust-libp2p/pull/5038#discussion_r1544928050 updated the code.
probably should have opened a new PR.... Ping @jxs how do you feel?

File: examples/metrics/src/http_service.rs [hunk 1]
```diff
@@ -18,45 +18,66 @@
...
 pub(crate) async fn metrics_server(registry: Registry) -> Result<(), std::io::Error> {
     // Serve on localhost.
-    let addr = ([127, 0, 0, 1], 0).into();
-
-    let server = Server::bind(&addr).serve(MakeMetricService::new(registry));
-    tracing::info!(metrics_server=%format!("http://{}/metrics", server.local_addr()));
-    if let Err(e) = server.await {
-        tracing::error!("server error: {}", e);
-    }
+    let addr: SocketAddr = ([127, 0, 0, 1], 0).into();
```
Comment (jxs) [inline examples/metrics/src/http_service.rs:36]: nit:
```suggestion
    let addr = ([127, 0, 0, 1], 0).into();
```
we can still keep the type elided right?

**Review by jxs — APPROVED** [review]
Thanks for this! Very nice work, LGTM!
(no worries regarding using this PR, I updated your comment to address the context, hope you don't mind)
CC @thomaseizinger since you started this

File: examples/metrics/src/http_service.rs (hunk 1 above)
Comment (drHuangMHT) [reply examples/metrics/src/http_service.rs:36]: No unfortunately.

**Comment by drHuangMHT** [comment]
Ping @thomaseizinger Got a minute to review this PR?

File: examples/metrics/src/http_service.rs [hunk 2]
```diff
@@ -18,45 +18,66 @@
...
+    let service = MetricService::new(registry);
+    let server = Router::new()
+        .route("/metrics", get(respond_with_metrics))
+        .fallback(respond_with_404_not_found)
```
Comment (thomaseizinger) [inline examples/metrics/src/http_service.rs]: I don't think we need an explicit fallback. Does `axum` not do that for us?

File: examples/metrics/src/http_service.rs [hunk 3]
```diff
@@ -18,45 +18,66 @@
...
+        .with_state(service);
+    let tcp_listener = TcpListener::bind(addr).await.unwrap();
```
Comment (thomaseizinger) [inline examples/metrics/src/http_service.rs]: Why not `?`?

```suggestion
    let tcp_listener = TcpListener::bind(addr).await?;
```

File: examples/metrics/src/http_service.rs [hunk 4]
```diff
@@ -18,45 +18,66 @@
...
     Ok(())
 }
 
+#[derive(Clone)]
 pub(crate) struct MetricService {
     reg: Arc<Mutex<Registry>>,
 }
 
+async fn respond_with_metrics(state: State<MetricService>) -> Response<String> {
+    state.respond_with_metrics()
```
Comment (thomaseizinger) [inline examples/metrics/src/http_service.rs]: This is a rather unusual route implementation I'd say. I'd expect `State` to just contain data and the have the functionality for making a response in this function here.

File: examples/metrics/src/http_service.rs [hunk 5]
```diff
@@ -18,45 +18,66 @@
...
+        Self {
+            reg: Arc::new(Mutex::new(registry)),
+        }
+    }
+
+    fn get_reg(&self) -> SharedRegistry {
         Arc::clone(&self.reg)
     }
-    fn respond_with_metrics(&mut self) -> Response<String> {
+    fn respond_with_metrics(&self) -> Response<String> {
```
Comment (thomaseizinger) [inline examples/metrics/src/http_service.rs]: We can simplify this with `axum`'s support for returning tuples: https://docs.rs/axum/latest/axum/response/index.html

File: misc/server/src/http_service.rs [hunk 6]
```diff
@@ -18,29 +18,47 @@
...
+    let server = Router::new()
+        .route(&metrics_path, get(respond_with_metrics))
+        .fallback(respond_with_404_not_found)
+        .with_state(service);
+    let tcp_listener = TcpListener::bind(addr).await?;
+    let local_addr = tcp_listener.local_addr()?;
+    tracing::info!(metrics_server=%format!("http://{}{}", local_addr, metrics_path));
+    axum::serve(tcp_listener, server.into_make_service())
+        .await
+        .unwrap();
```
Comment (thomaseizinger) [inline misc/server/src/http_service.rs]:
```suggestion
        .await?;
```

**Review by thomaseizinger — APPROVED** [review]
Did a quick pass, nothing blocking.

File: examples/metrics/src/http_service.rs (hunk 2 above)
Comment (drHuangMHT) [reply examples/metrics/src/http_service.rs]: Haven't worked with `axum` for a while now, I'm not sure about whether it provides default fallback. I simply direct that to existing implementation.

File: examples/metrics/src/http_service.rs (hunk 3 above)
Comment (drHuangMHT) [reply examples/metrics/src/http_service.rs]: Oops missed that

File: examples/metrics/src/http_service.rs (hunk 4 above)
Comment (drHuangMHT) [reply examples/metrics/src/http_service.rs]: Sure! I tried to make the change as small as possible to explore the possibility.

**Comment by thomaseizinger** [comment]
Can you solve the merge conflicts?

**Comment by drHuangMHT** [comment]
> Can you solve the merge conflicts?

Done. Staging this PR?

**Review by thomaseizinger — DISMISSED** [review]

File: examples/metrics/src/http_service.rs (hunk 2 above)
Comment (drHuangMHT) [reply examples/metrics/src/http_service.rs]: A quick search on `docs.rs` returned no info on default fallback. Not sure if `axum` provides such thing. Also we have a custom message for 404 response.

Comment (drHuangMHT) [reply examples/metrics/src/http_service.rs]: I ran the example `axum` setup and found that `axum` made minimal 404 NOT_FOUND response to routes that did not match anything. Is that acceptable?

Comment (thomaseizinger) [reply examples/metrics/src/http_service.rs]: Yeah that is totally fine. These are examples after all so I think less code is better. Users can make their own fallback if they decide to use this.

**Review by jxs — DISMISSED** [review]

**Comment by jxs** [comment]
Thanks @drHuangMHT  :heart:  and Thomas for the review

**Review by thomaseizinger — APPROVED** [review]

**Pull request merged**